olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/util/SubprocessConnection.java

```diff
@@ -192,112 +292,230 @@ public void setTimeout(int time, TimeUnit unit) {
-                if (process.isAlive()) {
-                    //
-                    // Announce that we're about to shut down.
-                    listeners.forEach(l -> l.subprocessPreShutdown(this));
-
-                    // First, tell it to shut down, then wait a short
-                    // period before making sure it is down.
-                    PrintWriter stdin = new PrintWriter(process.getOutputStream());
-                    stdin.println(SHUTDOWN);
-                    stdin.flush();
-                    try {
+        shutdown(true);
+    }
+
+    /**
+     * Shuts down the running subprocess, optionally giving the process a chance to
+     * close itself gracefully by sending it the text "SHUTDOWN" to on a line by itself.
+     * If the process is already running a command, this method will block until the
+     * process is done before shutting down. If you want a running command to shut
+     * down after a particular timeout, use {@link #setReadTimeout(int, TimeUnit)} to
+     * configure a timeout after which the process will be abandoned.
+     *
+     * @param graceful whether to shut down gracefully.
+     */
+    public void shutdown(boolean graceful) {
+        try {
+            processLock.lock();
+            if (process != null && process.isAlive()) {
+                //
+                // Announce that we're about to shut down.
+                listeners.forEach(l -> l.subprocessPreShutdown(this));
+
+                try {
+                    if (graceful) {
+                        // First, tell it to shut down, then wait a short
+                        // period before making sure it is down.
+                        PrintWriter stdin = new PrintWriter(process.getOutputStream());
+                        stdin.println(SHUTDOWN);
+                        stdin.flush();
                         if (!process.waitFor(5, TimeUnit.SECONDS)) {
                             process.destroyForcibly();
                         }
-                    } catch (InterruptedException e) {
-                        logger.log(Level.WARNING, "Shutdown interrupted", e);
+                    } else {
                         process.destroyForcibly();
                     }
-                    //
-                    // Announce that shutdown has happened.
-                    listeners.forEach(l -> l.subprocessPostShutdown(this));
+                } catch (InterruptedException e) {
+                    logger.log(Level.WARNING, "Shutdown interrupted", e);
+                    process.destroyForcibly();
+                } finally {
+                    process = null;
                 }
+                //
+                // Announce that shutdown has happened.
+                listeners.forEach(l -> l.subprocessPostShutdown(this));
+            } else {
+                //
+                // Announce that shutdown has happened.
+                listeners.forEach(l -> l.subprocessPostShutdown(this));
```

Comment:
This could be outside the if block and remove the else.

Reply:
Thanks - another leftover from an earlier bit of code that behaved a little differently in each case.